### PR TITLE
SDK-625 -- generate long link as fallback

### PR DESCRIPTION
**Pull request by apeterson-branch**
## TODO
* Test on Windows when the network is down.

## Reference
SDK-625 -- generate long link as fallback

## Description
Currently, if the internet connection is off - the SDK doesn't generate any links, it just returns an error.

## Testing Instructions
* Unit tests should continue to pass
* Samples should continue to work

I tested this by creating a unit test where -- with tracking disabled -- it generates a long link instead of a short one.

## Risk Assessment [`HIGH`]
* Note that there were a LOT of changes get this to work.   A lot of infrastructure didn't exist, and needed to be there.
* Note further that the Poco Base64 implementation had a series of poorly implemented APIs, and there are some hacks in here to get it to work correctly.

- [x] I, the PR creator, have tested — integration, unit, or otherwise — this code.

## Reviewer Checklist (To be checked off by the reviewer only)

- [ ] JIRA Ticket is referenced in PR title.
- Correctness & Style
    - [ ] Conforms to [Style Guides](https://google.github.io/styleguide/cppguide.html)
    - [ ] Mission critical pieces are documented in code and out of code as needed.
- [ ] Unit Tests reviewed and test issue sufficiently.
- [ ] Functionality was reviewed in QA independently by another engineer on the team.

### CODE REVIEW

File: BranchSDK/test/Util.h [hunk 1]
```diff
@@ -1,19 +1,67 @@
...
- * function.
- */
-void respondToRequest(BranchIO::IRequestCallback& callback)
-{
-    BranchIO::JSONObject::Ptr response(new BranchIO::JSONObject);
-    callback.onSuccess(0, response);
-}
-
-#endif // __UTIL_H__
+#define TEST_KEY "key_live_efTsR1fbTucbHvX3N5RsOaamDtlPFLap"
```
Comment (bklastaitis-branch) [inline BranchSDK/test/Util.h]: should this be a `key_test_xxx`?

File: BranchSDK/src/BranchIO/LinkInfo.cpp [hunk 2]
```diff
@@ -77,17 +146,95 @@ LinkInfo::setFeature(const std::string &feature) {
...
+
 LinkInfo &
 LinkInfo::setType(int type) {
     return doAddProperty(JSONKEY_TYPE, type);
 }
 
 
+
+void
+LinkInfo::generateUrl(Branch *branchInstance, IRequestCallback *callback) {
```
[the remark this answers is not in the document]
Comment (apeterson-branch) [reply BranchSDK/src/BranchIO/LinkInfo.cpp]: I'd like to keep both methods using the same verb -- as the user doesn't necessarily need to know implementation details.   With that said, I considered
* createUrl / createLongUrl
* getUrl / getLongUrl
* makeUrl / makeLongUrl

On Android it is `ServerRequestCreateUrl` -- but hides the ability to create a long Url only as a fallback to generating the request.

I will change this to use "create".

File: BranchSDK/test/Util.h (hunk 1 above)
Comment (apeterson-branch) [reply BranchSDK/test/Util.h]: Standardized usage of the test test key

File: BranchSDK/src/BranchIO/LinkInfo.cpp [hunk 3]
```diff
@@ -77,17 +146,95 @@ LinkInfo::setFeature(const std::string &feature) {
...
+        uri.addQueryParameter(JSONKEY_CAMPAIGN, campaign);
+    }
+
+    // Take the controlParams, convert those to a string, and base64 them
+    // Note that Poco Base64 introduces by default line endings after 74?? characters.  Call setLineLength(0) to fix.
+    // Note further that Poco Base64 implementation appears to require an eol, or it doesn't complete the encoding.
+    std::stringstream ss;
+    Poco::Base64Encoder b64enc(ss);
+    b64enc.rdbuf()->setLineLength(0);
+    b64enc << _controlParams.toString() << std::endl;
```
[the remark this answers is not in the document]
Comment (apeterson-branch) [reply BranchSDK/src/BranchIO/LinkInfo.cpp]: Done.

File: BranchSDK/src/BranchIO/JSONObject.cpp [hunk 4]
```diff
@@ -72,9 +72,15 @@ JSONObject::operator += (const JSONObject &rhs) {
...
 void
-JSONObject::set(const std::string& key, Ptr ptr) {
+JSONObject::set(const std::string& key, JSONObject::Ptr ptr) {
     Object& object(*ptr.get());
     Object::set(key, object);
 }
 
+void
+JSONObject::set(const std::string& key, Poco::JSON::Array::Ptr ptr) {
+    JSON::Array& array(*ptr.get());
```
[the remark this answers is not in the document]
Comment (apeterson-branch) [reply BranchSDK/src/BranchIO/JSONObject.cpp]: Addressed.

File: BranchSDK/src/BranchIO/LinkInfo.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (bklastaitis-branch) [reply BranchSDK/src/BranchIO/LinkInfo.cpp]: I was thinking more along the lines of, if _controlParams don't exist, what is the url even for? But we chatted in person and decided we should still support ability to create a link even if it all it does is open the app

[the remark this answers is not in the document]
Comment (apeterson-branch) [reply BranchSDK/src/BranchIO/LinkInfo.cpp]: I believe you can create a link without any control parameters.

**Pull request merged**